ee/hogai/trends/nodes.py

```diff
@@ -0,0 +1,381 @@
+            result = AgentAction("handle_incorrect_response", observation, text)
+
+        if isinstance(result, AgentFinish):
+            # Exceptional case
+            return {
+                "plan": result.log,
+                "intermediate_steps": None,
+            }
+
+        return {
+            "intermediate_steps": [*intermediate_steps, (result, None)],
+        }
+
+
+class CreateTrendsPlanToolsNode(AssistantNode):
+    name = AssistantNodeName.CREATE_TRENDS_PLAN_TOOLS
+
+    def router(self, state: AssistantState):
+        if state.get("plan") is not None:
+            return AssistantNodeName.GENERATE_TRENDS
+        return AssistantNodeName.CREATE_TRENDS_PLAN
+
+    def run(self, state: AssistantState, config: RunnableConfig):
+        toolkit = TrendsAgentToolkit(self._team)
+        intermediate_steps = state.get("intermediate_steps") or []
+        action, _ = intermediate_steps[-1]
+
+        try:
+            input = TrendsAgentToolModel.model_validate({"name": action.tool, "arguments": action.tool_input}).root
+        except ValidationError as e:
+            feedback = f"Invalid tool call. Pydantic exception: {e.errors(include_url=False)}"
+            return {"intermediate_steps": [*intermediate_steps, (action, feedback)]}
+
+        # The plan has been found. Move to the generation.
+        if input.name == "final_answer":
+            return {
+                "plan": input.arguments,
+                "intermediate_steps": None,
+            }
+
+        output = ""
+        if input.name == "retrieve_event_properties":
+            output = toolkit.retrieve_event_properties(input.arguments)
+        elif input.name == "retrieve_event_property_values":
+            output = toolkit.retrieve_event_property_values(input.arguments.event_name, input.arguments.property_name)
+        elif input.name == "retrieve_entity_properties":
+            output = toolkit.retrieve_entity_properties(input.arguments)
+        elif input.name == "retrieve_entity_property_values":
+            output = toolkit.retrieve_entity_property_values(input.arguments.entity, input.arguments.property_name)
+        else:
+            output = toolkit.handle_incorrect_response(input.arguments)
+
+        return {"intermediate_steps": [*intermediate_steps[:-1], (action, output)]}
+
+
+class GenerateTrendsNode(AssistantNode):
+    name = AssistantNodeName.GENERATE_TRENDS
+
+    @cached_property
+    def _group_mapping_prompt(self) -> str:
+        groups = GroupTypeMapping.objects.filter(team=self._team).order_by("group_type_index")
+        if not groups:
+            return "The user has not defined any groups."
+
+        root = ET.Element("list of defined groups")
+        root.text = (
+            "\n" + "\n".join([f'name "{group.group_type}", index {group.group_type_index}' for group in groups]) + "\n"
+        )
+        return ET.tostring(root, encoding="unicode")
```

Comment:
Why output XML?

Reply:
We output XML for data markup. I used lists that were also working, but XMLs or defined separators like XML tags seem to perform much better than any other formats. The trade-off, though, is the increased count of tokens.